pkg/reconciler/reconciler.go

```diff
@@ -167,6 +173,79 @@ func NewBase(opt Options, controllerAgentName string) *Base {
 	return base
 }
 
+// NewBase instantiates a new instance of Base implementing
+// the common & boilerplate code between our reconcilers.
+func NewInjectionBase(ctx context.Context, controllerAgentName string, cmw configmap.Watcher) *Base {
+	// Enrich the logs with controller name
+	logger := logging.FromContext(ctx).
+		Named(controllerAgentName).
+		With(zap.String(logkey.ControllerType, controllerAgentName))
+
+	kubeClient := kubeclient.Get(ctx)
+
+	recorder := GetEventRecorder(ctx)
+	if recorder == nil {
+		// Create event broadcaster
+		logger.Debug("Creating event broadcaster")
+		eventBroadcaster := record.NewBroadcaster()
+		watches := []watch.Interface{
+			eventBroadcaster.StartLogging(logger.Named("event-broadcaster").Infof),
+			eventBroadcaster.StartRecordingToSink(
+				&typedcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")}),
+		}
+		recorder = eventBroadcaster.NewRecorder(
+			scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
+		go func() {
+			<-ctx.Done()
+			for _, w := range watches {
+				w.Stop()
+			}
+		}()
+	}
+
+	statsReporter := GetStatsReporter(ctx)
+	if statsReporter == nil {
+		logger.Debug("Creating stats reporter")
+		var err error
+		statsReporter, err = NewStatsReporter(controllerAgentName)
+		if err != nil {
+			logger.Fatal(err)
+		}
+	}
+
+	base := &Base{
+		KubeClientSet:     kubeClient,
+		EventingClientSet: eventingclient.Get(ctx),
+		//ApiExtensionsClientSet: opt.ApiExtensionsClientSet,
+		DynamicClientSet: dynamicclient.Get(ctx),
+		ConfigMapWatcher: cmw,
+		Recorder:         recorder,
+		StatsReporter:    statsReporter,
+		Logger:           logger,
+	}
+
+	return base
+}
+
+// erKey is used to associate record.EventRecorders with contexts.
```

Comment:
Move all these EventRecorder functions be moved to their own file.

Reply:
Or better: knative/pkg (I wrote these in serving as well, and they are useful).

Reply:
I was wondering if there is potential reuse factor of some of this.